lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    for (const SILValue &escapingValue : escapingValues) {
+      if (DI->properlyDominates(escapingValue, src->getTerminator())) {
+        newArgs.push_back(escapingValue);
+      } else if (src != preheader) {
+        // newHeader arguments are available if src is not the preheader.
+        newArgs.push_back(newHeader->getArgument(argIndex));
+      } else {
+        // State from within the loop is not available in the preheader.
+        // Simply pass in an undef. This will never be accessed at runtime.
+        newArgs.push_back(getUndef(escapingValue->getType()));
+      }
+      ++argIndex;
+    }
+    // `exitIndex` to identify the block to which we exit from the loop.
+    // (insert a new value or get the old key value pair.)
+    auto emplaceResult = exitIndices.try_emplace(tgt, nextExitIndex);
+    if (emplaceResult.second) {
+      // Increment index as we inserted a new entry into the table.
+      ++nextExitIndex;
+    }
+    auto kvPair = *emplaceResult.first;
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 32,
+                                           /*exitIndex*/ kvPair.second));
+    // `stayInLoop` flag
+    newArgs.push_back(createTFIntegerConst(builder, location,
+                                           /*bitwidth*/ 1, stayInLoop));
+    appendArguments(src->getTerminator(), newTgt, newArgs);
+  }
+  return exitIndices;
+}
+
+/// Build a demuxing if..then..else block that can be used as a single
+/// exit block. This will not have any effect if there is a single exit
+/// block already.
+SILBasicBlock *SingleExitLoopTransformer::createNewExitBlock(
+    const llvm::DenseMap<SILBasicBlock *, intmax_t> &exitIndices,
+    SILValue exitIndexArg) {
+  auto createBlockOutsideLoop = [this]() {
+    SILBasicBlock *newBlock = currentFn->createBasicBlock();
+    SILLoop *parentLoop = loop->getParentLoop();
+    if (parentLoop) {
+      parentLoop->addBasicBlockToLoop(newBlock, LI->getBase());
+    }
+    return newBlock;
+  };
+
+  // Create a new exit block.
+  // FIXME: We can avoid creating an additional block and instead connect the
+  // header directly to the demuxBlock created in the loop below. Alternatively,
+  // we can also use contractUncondBranches in TFParititon.cpp to remove this
+  // block later.
+  SILBasicBlock *newExitBlock = createBlockOutsideLoop();
+
+  SILBuilder builder(newExitBlock);
+  ASTContext& context = builder.getASTContext();
+
+  auto curBlockIter = exitBlocks.rbegin();
+  SILBasicBlock *demuxBlock = *curBlockIter++;
+  SILLocation headerLocation =
+      getUserSourceLocation(header->getTerminator()->getDebugLocation());
+
+  while (curBlockIter != exitBlocks.rend()) {
+    SILBasicBlock *newBlock = createBlockOutsideLoop();
+    SILBasicBlock *trueBlock = *curBlockIter++;
+    builder.setInsertionPoint(newBlock);
+
+    // Create a condition to compare exitIndex to a constant
+    std::string equalOpName("Equal");
```

Comment:
this chunk of code looks similar to code in `handleSendRecvForTerminator()` in TFPartition.cpp. should we do some refactoring?